### ISSUE TITLE
Feature request: when importing midi files and they are type 0 then import after converting to type 1

### ISSUE THREAD

**Issue by hubertmeier** [helge17/tuxguitar#205]
Feature request: when importing midi files and they are **type 0** then import after converting to **type 1**

Often midi is delivered as type 0, alle notes in one track. Importing this gives only one track, the midi file has to be converted to type 1 (one track per instrument). I bought a midi-playback software and use only this conversion function ...

If this would be integrated ... that would be cool.

Thank you

**Comment by guiv42** [comment]
Could you please provide a simple example with the same song encoded as **type 0** and **type 1** ?

**Comment by hubertmeier** [comment]
Yes, my problem is, I need I licence free midi file, I can convert this in type 0 and type 1.

I buy these files at geerdes.com, for sure I can't provide them here ... They are copyrighted.

If you download a free sample at geerdes.com, you get usually type 0, all in one track.

Do you now a free sample track?

**Comment by guiv42** [comment]
I'm not very familiar with midi, to this should help if you can find some samples. In fact, I must admit I do not really understand your expectation.

Looking at the code, TuxGuitar is able to import both types 0 and 1 midi files.
But when it comes to exporting a file to midi, it seems to me it always exports as type 1 : [MidiFileWrite.write](https://github.com/helge17/tuxguitar/blob/30eea509753d2e9b61cef142421aae0d3cb324d2/common/TuxGuitar-midi/src/org/herac/tuxguitar/io/midi/MidiFileWriter.java#L14) is called with type==1 by [MidiSequenceHandlerImpl.notifyFinish](https://github.com/helge17/tuxguitar/blob/30eea509753d2e9b61cef142421aae0d3cb324d2/common/TuxGuitar-midi/src/org/herac/tuxguitar/io/midi/MidiSequenceHandlerImpl.java#L105).

So, except if there are other calls to MidiFileWriter I did not identify yet, I would say that successively calling "File/Import/Import Midi" then "File/Export/Export Midi" should convert an input midi file type 0 into an output midi file type 1.
If it does not work, try saving the file under .tg format after import, close it, re-open it, then export.

Could you have a try?

**Comment by hubertmeier** [comment]
OK, here i found a converter (Tool GN1:0) which can convert between 0 and 1 and the other way:

**https://www.gnmidi.com/gnfreeen.htm**

it works, i tested it. There is a demo in the download, the delivered demo has format 1, Tuxguitar import this correct. If i convert to type 0 which is often delivered, Tuxguitar imports only one track what is correct - the format 0 has only one track - and mixes all notes from all tracks into one track.
I think, type 1 has notes which belongs to a track and the track has its assigned instrument. Format 0 has only one track and every note is assigned to his instrument directly which is ignored by tuxguitar, but i have not much knowledges about midi formats.

Another issue is that tuxguitar imports type 1 correct but often there are no names for the tracks, normally is the name of the track the assigned instrument. I saw this behavier since some versions, i think tg 1.2 did not have this problem.

**Comment by guiv42** [comment]
I'm getting a little confused, so let me try to rephrase your questions to make sure I understand correctly.
You have 2 distinct problems.

1. midi type 0

When importing a midi type 0 file (so with 1 single midi track) with notes played by multiple instruments, you expect TuxGuitar to split song into multiple tracks and it doesn't. Is my interpretation correct?
I cannot test this scenario since I don't have any such file, and I can't use GN1:0 since it's a windows-only tool. Here is a basic sample file under TuxGuitar and midi type 1 formats.
[test.zip](https://github.com/helge17/tuxguitar/files/13942230/test.zip)

Could you please, with GN1:0
- convert midi file from type 1 to type 0
- convert *back* from type 0 to type 1
then attach the 2 files under zip format here?


2. Tracks names

>  tuxguitar imports type 1 correct but often there are no names for the tracks, normally is the name of the track the assigned instrument

From my understanding of midi type 1, tracks names should be defined *in the midi file* with "Track Name" events (FF 03). It seems to me these events are correctly [imported](https://github.com/helge17/tuxguitar/blob/271138fcfb0a9e2c55533b1c85add22841a17d8d/common/TuxGuitar-midi/src/org/herac/tuxguitar/io/midi/MidiSongReader.java#L259C10-L259C10) by TuxGuitar. There is a priori no reason for a track to be named with its instrument. If you have empty track names, I'd say most probably the tracks have no name in your midi file.
See the attached files above: when I export a TuxGuitar file under midi (type 1) then re-import it, tracks names are OK.
In this case it would mean your problem is not a TuxGuitar issue.

**Comment by hubertmeier** [comment]
First of all, thank you for the time you are taking to make this request. The problem is not that important, these converters exist. Still, it would be cool to have this in Tuxguitar.

Your description is exactly right, an import from Type 0 mixes all notes onto one track, new tracks should have to be created for notes that represent another instrument and the notes and perhaps other midi events should be distributed across these tracks.

The zip-file contains your test converted with two different programs, GN0 and Synthfont. The results are the same.

[test_type_0.zip](https://github.com/helge17/tuxguitar/files/13952278/test_type_0.zip)

![type 0 imported](https://github.com/helge17/tuxguitar/assets/20687189/cc1b723e-8dfd-4ce7-8e49-2e02c1ebb08b)

thank you

**Comment by guiv42** [comment]
I'm sorry but I'll be fully sincere here: I won't even read this [AI-slop](https://en.wikipedia.org/wiki/AI_slop)
Pushing the output of an AI here without any further analysis is just pushing an additional burden on the shoulders of the maintainers. Zero added value.
Just don't do that.

**Comment by alexisguerreroabg** [comment]
> I'm sorry but I'll be fully sincere here: I won't even read this [AI-slop](https://en.wikipedia.org/wiki/AI_slop) Pushing the output of an AI here without any further analysis is just pushing an additional burden on the shoulders of the maintainers. Zero added value. Just don't do that.

Sorry, I removed the comment.

**Comment by guiv42** [comment]
Thanks.
